std/math/big/int.zig

```diff
@@ -0,0 +1,2023 @@
+        var a = wrapInt(&stack.allocator, av);
+        var b = wrapInt(&stack.allocator, bv);
+
+        if (a.positive != b.positive) {
+            if (a.positive) {
+                // (a) - (-b) => a + b
+                const bp = Int{
+                    .allocator = undefined,
+                    .positive = true,
+                    .limbs = b.limbs,
+                    .len = b.len,
+                };
+                try r.add(a, bp);
+            } else {
+                // (-a) - (b) => -(a + b)
+                const ap = Int{
+                    .allocator = undefined,
+                    .positive = true,
+                    .limbs = a.limbs,
+                    .len = a.len,
+                };
+                try r.add(ap, b);
+                r.positive = false;
+            }
+        } else {
+            if (a.positive) {
+                // (a) - (b) => a - b
+                if (a.cmp(b) >= 0) {
+                    try r.ensureCapacity(a.len + 1);
+                    llsub(r.limbs[0..], a.limbs[0..a.len], b.limbs[0..b.len]);
+                    r.normN(a.len);
+                    r.positive = true;
+                } else {
+                    try r.ensureCapacity(b.len + 1);
+                    llsub(r.limbs[0..], b.limbs[0..b.len], a.limbs[0..a.len]);
+                    r.normN(b.len);
+                    r.positive = false;
+                }
+            } else {
+                // (-a) - (-b) => -(a - b)
+                if (a.cmp(b) < 0) {
+                    try r.ensureCapacity(a.len + 1);
+                    llsub(r.limbs[0..], a.limbs[0..a.len], b.limbs[0..b.len]);
+                    r.normN(a.len);
+                    r.positive = false;
+                } else {
+                    try r.ensureCapacity(b.len + 1);
+                    llsub(r.limbs[0..], b.limbs[0..b.len], a.limbs[0..a.len]);
+                    r.normN(b.len);
+                    r.positive = true;
+                }
+            }
+        }
+    }
+
+    // Knuth 4.3.1, Algorithm S.
+    fn llsub(r: []Limb, a: []const Limb, b: []const Limb) void {
+        @setRuntimeSafety(false);
+        debug.assert(a.len != 0 and b.len != 0);
+        debug.assert(a.len > b.len or (a.len == b.len and a[a.len - 1] >= b[b.len - 1]));
+        debug.assert(r.len >= a.len);
+
+        var i: usize = 0;
+        var borrow: Limb = 0;
+
+        while (i < b.len) : (i += 1) {
+            var c: Limb = 0;
+            c += Limb(@subWithOverflow(Limb, a[i], b[i], &r[i]));
+            c += Limb(@subWithOverflow(Limb, r[i], borrow, &r[i]));
+            borrow = c;
+        }
+
+        while (i < a.len) : (i += 1) {
+            borrow = Limb(@subWithOverflow(Limb, a[i], borrow, &r[i]));
+        }
+
+        debug.assert(borrow == 0);
+    }
+
+    // rma = a * b
+    //
+    // For greatest efficiency, ensure rma does not alias a or b.
+    pub fn mul(rma: *Int, av: var, bv: var) !void {
```

Comment:
I have actually implemented Karatsuba (and Toom-3) variants for this, but they were a bit unclean to implement well (no low-level limbs with temporary allocations) so I haven't pushed them anywhere just yet. These are only useful for balanced operand sizes as well, so I expect the main use case right now wouldn't get any benefits. Unbalanced multiplication requires some other algorithms which I haven't explored in detail.